sycl/include/sycl/ext/intel/experimental/esimd/detail/simd_view_impl.hpp

```diff
@@ -193,6 +193,8 @@ template <typename BaseTy, typename RegionTy> class simd_view_impl {
   }
 
 #define DEF_BINOP(BINOP, OPASSIGN)                                             \
+  template <class T1 = simd_view_impl,                                         \
```

Comment:
As I understood `a[0] + x` was ambiguous when the underlying types were different, right? E.g., `int` and `uint32_t`.
You resolve it by disabling one of the candidates and rely on implicit conversion to underlying type and built-in operator+ for primitive types.

Can we sink those two BINOP overloads into the derived class of generic `simd_view` implementation? That way I think we don't need this SFINAE trick.

Reply:
> As I understood a[0] + x was ambiguous when the underlying types were different, right? E.g., int and uint32_t.

not really. int - int causes the failure too. In the test I just modelled the user test case where the types were different.

> You resolve it by disabling one of the candidates and rely on implicit conversion to underlying type and built-in operator+ for primitive types.

yes, with length == 1, simd_view_impl and value_type are implicitly convertible to the element type, thus compiler will always be able to resolve `BINOP(const simd_view_impl &, const value_type &)` to the built-in operation over primitive types after implicitly converting both sides.

> Can we sink those two BINOP overloads into the derived class of generic simd_view implementation?

not sure I understand. Do you mean moving BINOP from simd_view_impl to simd_view? That won't affect resolution in any way, since simd_view already binds to simd_view_impl w/o any conversions as simd_view extends simd_view_impl.

Reply:
Per discussion with @DenisBakhvalov: his suggestion was to move BINOP to simd_view to make them unavailable for the length=1 simd_view specialization. That might make sense, but (1) would be pretty intrusive change at this point (2) my upcoming simd_mask introduction patch refactors operators anyway (none will be friends), so this can be re-visited when reviewing that patch. So we decided to go with the current fix at this point.